### PR TITLE
fix: [PL-42443]: enable the disabled user when they are added to new accounts

### PR DESCRIPTION
**Pull request by sahibaHarness**
## Describe your changes
When an account is disabled if a user is not part of any more accounts the user as well is marked disabled.
This is to enable the user when it's being invited to a new account

Add user flow is common while added a user to an account.

Tested for:
-> Added user in CG 
-> Added user in NG 

https://github.com/harness/harness-core/assets/112424209/ef82f304-2dde-4f98-85d3-66de0cbc2265

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55620)

### CODE REVIEW

**Comment by sahibaHarness** [comment]
trigger messagecheck

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55620)* status: 1 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/UserServiceImpl.java` line 4538 at r1](https://reviewable.io/reviews/harness/harness-core/55620#-NjfiNOh2HCJ-oCN5lQF:-NjfiNOh2HCJ-oCN5lQG:bdo2bwl) ([raw file](https://github.com/harness/harness-core/blob/618af33b1f3a5bcca3319a6fc73cd9e5550570bc/400-rest/src/main/java/software/wings/service/impl/UserServiceImpl.java#L4538)):*
> ```Java
>     if (existingUser.getDisabled()) {
>       updateOperations.set(UserKeys.disabled, false);
>     }
> ```

We should add UTs for this and other change.

File: 400-rest/src/main/java/software/wings/service/impl/UserServiceImpl.java [hunk 1]
```diff
@@ -1066,6 +1066,9 @@ public void addUserToAccount(String userId, String accountId, UserSource userSou
       updateOperations.set(UserKeys.accounts, newAccounts);
     }
 
+    if (user.isDisabled()) {
+      updateOperations.set(UserKeys.disabled, false);
+    }
```
Comment (pbarapatre10) [inline 400-rest/src/main/java/software/wings/service/impl/UserServiceImpl.java:1071]: We should also change the default account of user to the new account where it is set to the disabled account earlier. Consider the scenario: User is disabled, with its default account been set to the disabled account. Now when user is added to new account its default account would still be the disabled account and login most likely will not work, despite now the user being part of the new account. Please handle the scenario and test. Thanks

File: 400-rest/src/main/java/software/wings/service/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -4530,6 +4534,10 @@ private void moveAccountFromPendingToConfirmed(
     updateOperations.set(UserKeys.accounts, newAccountsList);
     updateOperations.set(UserKeys.pendingAccounts, newPendingAccountsList);
     updateOperations.set(UserKeys.emailVerified, markEmailVerified);
+    if (existingUser.getDisabled()) {
+      updateOperations.set(UserKeys.disabled, false);
+      updateOperations.set(UserKeys.defaultAccountId, invitationAccount);
```
Comment (pbarapatre10) [inline 400-rest/src/main/java/software/wings/service/impl/UserServiceImpl.java:4539]: here it should be set to `invitationAccount.getUuid()` , thanks

**Comment by sahibaHarness** [comment]
trigger uts

**Review by pbarapatre10 — APPROVED** [review]

**Review by adiyaar24 — APPROVED** [review]

**Comment by sahibaHarness** [comment]
trigger uts

**Comment by sahibaHarness** [comment]
trigger ut5

**Pull request merged**